support/Pipelines/Bbh/ControlId.py

```diff
@@ -0,0 +1,173 @@
+                "\n"
+                "=========================================="
+                f" Control of BBH Parameters ({iteration}) "
+                "=========================================="
+            )
+            control_run_dir = f"{id_run_dir}/ControlParams{iteration:02}"
+
+            # Run ID and find horizons
+            generate_id(
+                mass_a=M_Kerr_A,
+                mass_b=M_Kerr_B,
+                dimensionless_spin_a=[chi_Kerr_A_x, chi_Kerr_A_y, chi_Kerr_A_z],
+                dimensionless_spin_b=[chi_Kerr_B_x, chi_Kerr_B_y, chi_Kerr_B_z],
+                separation=separation,
+                orbital_angular_velocity=orbital_angular_velocity,
+                radial_expansion_velocity=radial_expansion_velocity,
+                run_dir=control_run_dir,
+                scheduler=None,
+            )
+
+        # Get black hole physical parameters
+        horizons_file = h5py.File(f"{control_run_dir}/Horizons.h5", "r")
+        M_A = horizons_file["AhA.dat"][0, 4]  # ChristodoulouMass
+        M_B = horizons_file["AhB.dat"][0, 4]  # ChristodoulouMass
+        chi_A_x = horizons_file["AhA.dat"][0, 6]  # DimensionlessSpinVector_x
+        chi_A_y = horizons_file["AhA.dat"][0, 7]  # DimensionlessSpinVector_y
+        chi_A_z = horizons_file["AhA.dat"][0, 8]  # DimensionlessSpinVector_z
+        chi_B_x = horizons_file["AhB.dat"][0, 6]  # DimensionlessSpinVector_x
+        chi_B_y = horizons_file["AhB.dat"][0, 7]  # DimensionlessSpinVector_y
+        chi_B_z = horizons_file["AhB.dat"][0, 8]  # DimensionlessSpinVector_z
+
+        residual = np.array(
+            [
+                M_A - M_target_A,
+                M_B - M_target_B,
+                chi_A_x - chi_target_A[0],
+                chi_A_y - chi_target_A[1],
+                chi_A_z - chi_target_A[2],
+                chi_B_x - chi_target_B[0],
+                chi_B_y - chi_target_B[1],
+                chi_B_z - chi_target_B[2],
+            ]
+        )
+        print("Control Residual =", residual)
+
+        data_file.write(
+            f" {iteration},"
+            f" {M_A}, {M_B},"
+            f" {chi_A_x}, {chi_A_y}, {chi_A_z},"
+            f" {chi_B_x}, {chi_B_y}, {chi_B_z},"
+            f" {residual[0]}, {residual[1]},"
+            f" {residual[2]}, {residual[3]}, {residual[4]},"
+            f" {residual[5]}, {residual[6]}, {residual[7]}"
+            " \n"
+        )
+        data_file.flush()
+
+        return residual
+
+    # Initial guess
+    u = np.array([M_target_A, M_target_B, *chi_target_A, *chi_target_B])
+
+    # Initial residual
+    F = Residual(*u)
+
+    # Initial Jacobian (identity matrix)
+    J = np.identity(len(u))
+
+    while iteration < max_iterations and np.max(np.abs(F)) > residual_tolerance:
+        iteration += 1
+        u_prev = u
+
+        # Update the free parameters using a quasi-Newton-Raphson method
+        u = u_prev - np.dot(np.linalg.inv(J), F)
+
+        Delta_u = u - u_prev
```

Comment:
I would find it easier to understand if you do this:

- Move this `delta_u` two lines up, and set it to `delta_u = np.dot(np.linalg.inv(J), F)`
- In your `residual` function, add a minus sign so you return `target - measured`.
- Then here you can do `u += delta_u` and you don't need the variable `u_prev` anymore.